samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,323 @@
  */
 package org.apache.samza.operators.impl;
 
-import org.apache.samza.config.Config;
-import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.data.MessageEnvelope;
 import org.apache.samza.operators.spec.WindowOperatorSpec;
+import org.apache.samza.operators.triggers.Cancellable;
+import org.apache.samza.operators.triggers.RepeatingTriggerImpl;
+import org.apache.samza.operators.triggers.TimeTrigger;
+import org.apache.samza.operators.triggers.Trigger;
+import org.apache.samza.operators.triggers.TriggerContext;
+import org.apache.samza.operators.triggers.TriggerImpl;
+import org.apache.samza.operators.triggers.TriggerImpls;
+import org.apache.samza.operators.util.InternalInMemoryStore;
+import org.apache.samza.operators.windows.AccumulationMode;
+import org.apache.samza.operators.windows.WindowKey;
 import org.apache.samza.operators.windows.WindowPane;
 import org.apache.samza.operators.windows.internal.WindowInternal;
+import org.apache.samza.operators.windows.internal.WindowType;
+import org.apache.samza.storage.kv.KeyValueStore;
 import org.apache.samza.task.MessageCollector;
-import org.apache.samza.task.TaskContext;
 import org.apache.samza.task.TaskCoordinator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-public class WindowOperatorImpl<M, WK, WV> extends OperatorImpl<M, WindowPane<WK, WV>> {
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.function.BiFunction;
+import java.util.function.Function;
 
-  private final WindowInternal<M, WK, WV> window;
+/**
+ * Implementation of a window operator that groups messages into finite windows for processing.
+ *
+ * <p>Notes:
+ * This class implements the processing logic for various types of windows and triggers. It tracks and manages state for
+ * all open windows, the active triggers that correspond to each of the windows and the pending callbacks. It provides
+ * an implementation of {@link TriggerContext} that {@link TriggerImpl}s can use to schedule and cancel callbacks. It
+ * also orchestrates the flow of messages through the various {@link TriggerImpl}s.
+ *
+ * @param <M>  the type of the incoming {@link MessageEnvelope}
+ * @param <K>  the type of the key in this {@link org.apache.samza.operators.MessageStream}
+ * @param <WK> the type of the key in the emitted window pane
+ * @param <WV> the type of the value in the emitted window pane
+ * @param <WM> the type of the emitted window pane
+ *
+ * TODO: Implement expiration of entries and triggers
+ */
+public class WindowOperatorImpl<M extends MessageEnvelope, K, WK, WV, WM extends WindowPane<K, WV>> extends OperatorImpl<M, WM> {
```

Comment:
Good suggestion, a nice fallout is that there's one less type parameter in the `WindowOperatorImpl` (which I <3).